src/fdc/MSXtar.cc

```diff
@@ -194,38 +212,48 @@ MSXtar::~MSXtar()
 	}
 }
 
-// transform BAD_FAT (0xFF7) and EOF_FAT-range (0xFF8-0xFFF)
-// to a single value: EOF_FAT (0xFFF)
-static constexpr unsigned normalizeFAT(unsigned cluster)
-{
-	return (cluster < BAD_FAT) ? cluster : EOF_FAT;
-}
-
 // Get the next cluster number from the FAT chain
+// BAD FAT and EOF FAT-range are normalized to a single EOF_FAT value (0xFFFFFFF)
 unsigned MSXtar::readFAT(unsigned clNr) const
 {
 	assert(!fatBuffer.empty()); // FAT must already be cached
 	std::span<const uint8_t> data{fatBuffer[0].raw.data(), sectorsPerFat * size_t(SECTOR_SIZE)};
-	auto p = subspan<2>(data, (clNr * 3) / 2);
-	unsigned result = (clNr & 1)
-	                ? (p[0] >> 4) + (p[1] << 4)
-	                : p[0] + ((p[1] & 0x0F) << 8);
-	return normalizeFAT(result);
+	if (fat16) {
+		auto p = subspan<2>(data, clNr * 2);
+		unsigned result = p[0] | p[1] << 8;
+		return result < BAD_FAT16 ? result : EOF_FAT;
+	} else {
+		auto p = subspan<2>(data, (clNr * 3) / 2);
+		unsigned result = (clNr & 1)
+						? (p[0] >> 4) + (p[1] << 4)
+						: p[0] + ((p[1] & 0x0F) << 8);
+		return result < BAD_FAT12 ? result : EOF_FAT;
+	}
 }
 
 // Write an entry to the FAT
 void MSXtar::writeFAT(unsigned clNr, unsigned val)
 {
 	assert(!fatBuffer.empty()); // FAT must already be cached
-	assert(val < 4096); // FAT12
-	std::span data{fatBuffer[0].raw.data(), sectorsPerFat * size_t(SECTOR_SIZE)};
-	auto p = subspan<2>(data, (clNr * 3) / 2);
-	if (clNr & 1) {
-		p[0] = narrow_cast<uint8_t>((p[0] & 0x0F) + (val << 4));
-		p[1] = narrow_cast<uint8_t>(val >> 4);
+	if (val == EOF_FAT) {
+		val = fat16 ? EOF_FAT16 : EOF_FAT12;
 	} else {
+		assert(fat16 ? val < MAX_FAT16 : val < MAX_FAT12);
```

Comment:
Detail. This confused me for a second. Maybe rewrite as:
`assert(val < (fat16 ? MAX_FAT16 : MAX_FAT12));`
This check implies `MAX_FAT` is exclusive (see above), is that correct?

Reply:
Yes, cluster number 0xFF6 / 0xFFF6 is [reserved](https://en.wikipedia.org/wiki/Design_of_the_FAT_file_system#cite_ref-Ecma-107_5-8).